File: gtci/17-modified-binary-search/python3/7-bitonic-array-maximum.py
class Solution:
    def findMax(self, arr):
        if not arr:
            return -1

        left = 0
        right = len(arr) - 1

        # instead of comparing against target, we will work out
        # whether we are in the increasing half or the decreasing half
        while left < right:
            mid = (left + right) // 2

            if mid < len(arr) - 1 and arr[mid] < arr[mid + 1]:
                # increasing half, so converge towards the peak
                left = mid + 1
            else:
                # decreasing half, so converge backwards to the peak
                right = mid

        return arr[left]


if __name__ == "__main__":
    sol = Solution()

    assert sol.findMax([1, 3, 8, 4, 3]) == 8
    assert sol.findMax([10, 20, 30, 40, 50]) == 50
    assert sol.findMax([120, 100, 80, 20, 0]) == 120
    assert sol.findMax([]) == -1

    print("All test cases passed.")

File: gtci/17-modified-binary-search/cpp/7-bitonic-array-maximum.cpp
#include <cassert>
#include <iostream>
#include <vector>

class Solution {
public:
  int findMax(const std::vector<int>& arr) {
    if (arr.empty())
      return -1;
    
    int left = 0;
    int right = arr.size() - 1;

    while (left < right) {
      const int mid = left + (right - left) / 2;

      if (mid < arr.size() - 1 && arr[mid] < arr[mid + 1])
        left = mid + 1;
      else
        right = mid;
    }

    return arr[left];
  }
};

int main() {
    Solution sol;

    assert(sol.findMax({1, 3, 8, 12, 4, 2}) == 12);
    assert(sol.findMax({3, 8, 3, 1}) == 8);
    assert(sol.findMax({1, 3, 8, 12}) == 12);
    assert(sol.findMax({10, 9, 8}) == 10);
    assert(sol.findMax({10}) == 10);
    assert(sol.findMax({}) == -1);

    std::cout << "All test cases passed." << std::endl;

}